lib/vega_lite/data.ex

```diff
@@ -0,0 +1,231 @@
+  `:text` field. Defaults to `:nominal` for the axes and `:quantitative` for color and text if
+  types are not specified.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+
+  """
+  def heatmap(data, fields), do: heatmap(Vl.new(), data, fields)
+
+  @doc """
+  Same as heatmap/2, but takes a valid `VegaLite` specification as the first argument.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Vl.new(title: "Heatmap", width: 500)
+      |> Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+  """
+  def heatmap(vl, data, fields) do
+    fields = Enum.map(fields, &heatmap_defaults/1)
+
+    if fields[:text],
+      do: annotated_heatmap(vl, data, fields),
+      else: chart(vl, data, :rect, fields)
+  end
+
+  defp annotated_heatmap(vl, data, fields) do
+    text_fields = [text: fields[:text], x: fields[:x], y: fields[:y]]
+    fields = List.keydelete(fields, :text, 0)
+    used_fields = fields |> Keyword.values() |> used_fields()
+
+    vl
+    |> Vl.data_from_values(data, only: used_fields)
+    |> Vl.layers([layer(data, :rect, fields), layer(data, :text, text_fields)])
+  end
+
+  defp heatmap_defaults({field, [{:field, _col} | _] = opts}) when field in [:x, :y] do
+    {field, Keyword.put_new(opts, :type, :nominal)}
+  end
+
+  defp heatmap_defaults({field, col}) when field in [:x, :y] do
+    {field, [field: col, type: :nominal]}
+  end
+
+  defp heatmap_defaults({field, [{:field, _col} | _] = opts}) when field in [:color, :text] do
+    {field, Keyword.put_new(opts, :type, :quantitative)}
+  end
+
+  defp heatmap_defaults({field, col}) when field in [:color, :text] do
+    {field, [field: col, type: :quantitative]}
+  end
+
+  defp enconde_mark(vl, [{:type, mark} | opts]), do: Vl.mark(vl, mark, opts)
+  defp enconde_mark(vl, mark), do: Vl.mark(vl, mark)
+
+  defp encode_field(schema, cols, field, [{:field, col} | opts]) do
+    opts = Keyword.put_new(opts, :type, type_for(cols, col))
+    Vl.encode_field(schema, field, col, opts)
+  end
+
+  defp encode_field(schema, cols, field, col) do
+    Vl.encode_field(schema, field, col, type: type_for(cols, col))
+  end
+
+  defp layer(data, mark, fields) do
+    cols = columns_for(data)
+    root = Vl.new() |> enconde_mark(mark)
+
+    for {field, col} <- fields, reduce: root do
+      acc -> encode_field(acc, cols, field, col)
+    end
+  end
+
+  defp used_fields(fields) do
+    for field <- fields, field, uniq: true do
+      if is_list(field), do: field[:field], else: field
+    end
+  end
+
+  defp type_for(data, name), do: get_in(data, [Access.filter(&(&1.name == name)), :type]) |> hd()
+
+  defp columns_for(data) do
```

Comment:
I am almost sure we have this logic elsewhere, don't we? Maybe it is on Kino.VegaLite though? Maybe we should expose this here and reuse there? Something like `VegaLite.Data.infer_types(...)`?

Reply:
With minimal differences, but yes!